src/couch_views/src/couch_views_indexer.erl

```diff
@@ -165,48 +171,60 @@ add_error(Error, Reason, Data) ->
 
 
 update(#{} = Db, Mrst0, State0) ->
-    Limiter = maps:get(limiter, State0),
-    case couch_rate:budget(Limiter) of
-        0 ->
-            couch_rate:wait(Limiter),
-            update(Db, Mrst0, State0);
-        Limit ->
-            {Mrst1, State1} = do_update(Db, Mrst0, State0#{limit => Limit, limiter => Limiter}),
-            case State1 of
-                finished ->
-                    couch_eval:release_map_context(Mrst1#mrst.qserver);
-                _ ->
-                    couch_rate:wait(Limiter),
-                    update(Db, Mrst1, State1)
+    Limit = couch_views_batch:start(),
+    try
+        {Mrst1, State1} = do_update(Db, Mrst0, State0#{limit => Limit}),
+        case State1 of
+            finished ->
+                couch_eval:release_map_context(Mrst1#mrst.qserver);
+            _ ->
+                #{
+                    docs_read := DocsRead,
+                    tx_size := TxSize
+                } = State1,
+                couch_views_batch:success(TxSize, DocsRead),
+                update(Db, Mrst1, State1)
```

Comment:
Because of the catch that wraps it, would this still be tail recursive? Say if we call this 1M times would it have to keep the stack frames around to unwind after the call?

Reply:
Ooooh, good catch. Will fix that.